authorization/storage.go

```diff
@@ -2,33 +2,320 @@ package authorization
-		IDGen:   snowflake.NewDefaultIDGenerator(),
+// autogenerateHasher generates an AuthorizationHasher that hashes using variantName.
+// The decoders include variantName plus any other variants that are detected in the
+// store.
+func (s *Store) autogenerateHasher(ctx context.Context, variantName string) (*AuthorizationHasher, error) {
+	// Determine which variants are present in the store.
+	tempDecoder := crypt.NewDecoder()
+	if err := influxdb2_algo.RegisterDecoder(tempDecoder); err != nil {
+		return nil, fmt.Errorf("error registering test decoders for authorization store: %w", err)
 	}
-	return st, st.setup()
+
+	var auths []*influxdb.Authorization
+	err := s.View(ctx, func(tx kv.Tx) error {
+		as, err := s.ListAuthorizations(ctx, tx, influxdb.AuthorizationFilter{})
+		if err != nil {
+			return err
+		}
+		auths = as
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error fetching authorization records for hash variant inventory: %w", err)
+	}
+
+	foundVariants := make(map[influxdb2_algo.Variant]struct{})
+	for _, a := range auths {
+		if a.HashedToken != "" {
+			digest, err := tempDecoder.Decode(a.HashedToken)
+			if err == nil {
+				if influxdbDigest, ok := digest.(*influxdb2_algo.Digest); ok {
+					foundVariants[influxdbDigest.Variant] = struct{}{}
+				}
+			}
+		}
+	}
+
+	var decoderVariants []influxdb2_algo.Variant
+	// Make sure we have the hasher variant we will make in there and that it is first in the list,
+	// so that it is the first one we try to lookup a given token.
+	hasherVariant := influxdb2_algo.NewVariant(variantName)
+	decoderVariants = append(decoderVariants, hasherVariant)
+	for variant := range foundVariants {
+		// Avoid having 2 hasherVariant decoders.
+		if variant != hasherVariant {
+			decoderVariants = append(decoderVariants, variant)
+		}
+	}
+
+	hasher, err := NewAuthorizationHasher(WithHasherVariant(hasherVariant), WithDecoderVariants(decoderVariants))
+	if err != nil {
+		return nil, fmt.Errorf("error creating authorization hasher for authorization store: %w", err)
+	}
+
+	return hasher, nil
+}
+
+// hashedTokenMigration migrates any unhashed tokens in the store to hashed tokens.
+func (s *Store) hashedTokenMigration(ctx context.Context) error {
+	if !s.useHashedTokens || s.readOnly {
+		return nil
+	}
+
+	// Figure out which authorization records need to be updated.
+	var authsNeedingUpdate []*influxdb.Authorization
+	err := s.View(ctx, func(tx kv.Tx) error {
+		s.forEachAuthorization(ctx, tx, nil, func(a *influxdb.Authorization) bool {
+			if a.HashedToken == "" {
+				if a.Token != "" {
+					authsNeedingUpdate = append(authsNeedingUpdate, a)
+				} else {
+					s.log.Warn("found authorization without any token set during hashed token migration", zap.Uint64("ID", uint64(a.ID)), zap.String("description", a.Description))
+				}
+			}
+			return true
+		})
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	for batch := range slices.Chunk(authsNeedingUpdate, 100) {
```

Comment:
AI is concerned that the getting of tokens to update and the updating of them aren't both done in a streaming fashion. `authsNeedingUpdate` accumulates all the tokens. I told AI that 1 million tokens is highly unlikely and then it said 

> I concede the migration performance concern is overstated.